### ISSUE TITLE
Access each containers by easy understandable name with Docker overlay driver in Swarm Mode

### ISSUE THREAD

**Issue by nicodmf** [moby/moby#30546]
Not a bug, a feature, the facility to access a host by service.id in the dns tools.

 #Example :  with a service mongo with 3 replicas : 

Each replica should be accessible by : 
- {servicename}.{replica.id}                           
   - ex: mongo.1 
- {servicename}.{nodename}                         
  - ex: mongo.node1
- {servicename}.{nodename}.{replica.id}                        
  - ex: mongo.node1.1
- {nodename}.{servicename} 
  - ex: node1.mongo
- {nodename}.{servicename}.{replica.id} 
  - ex: node1.mongo.1
- {serviceid} 
  - ex: g1xlvn56l30k

https://github.com/docker/libnetwork/pull/1372
https://github.com/docker/docker/pull/25420

**Comment by thaJeztah** [comment]
can you explain the exact use case? I think the problem here may be that tasks get replaced when a service gets updated; in that case, old containers are stopped, but not removed

**Comment by nicodmf** [comment]
Many use case but the first is, i wanted create juste one mongodb container with an entrypoint capable to start each container with different roles (a master, 2 secondary), actually i should create 3 containers creating a big différence between my docker-compose.yml for dev and for prod.

Another case is, i have a container client/console on each swarm node. If i connnect to this console container, when i would operate on another container (app, db, or another) with an easy way and a clearly name.

**Comment by aluzzardi** [comment]
I think we talked about this in the past and I think it would be a nice feature to have.

/cc @mavenugo

**Comment by udiromano** [comment]
Any idea when this feature will be available?

**Comment by mrapczynski** [comment]
This would be awesome to have.

I have a use case where I am deploying an etcd cluster as a session store for a web frontend, and I need to define etcd each node as a separate service.

If each container/replica/task in a service had an addressable name, then I would be able to simplify the deployment to just one service and _n_ replicas. Etcd needs a host list provided over an environment variable such as `ETCD_INITIAL_CLUSTER=shib-etcd0=http://etcd0:2380,shib-etcd1=http://etcd1:2380`, and each node needs a stable name.

**Comment by ksachdev1** [comment]
The first message in the thread points to docker/libnetwork#1372
#25420
Is this resolved now? I am still trying to solve the following in swarm mode, does it work now, if yes can someone point me to the documentation?
I have a docker swarm on which say I deploy multiple instances of MongoDB. The instances on doing docker service ls services_mongo will show something along these lines

service_mongo.1
service_mongo.2
service_mongo.3
My web services need individually pingable names for the mongo service and the url for mongo is composed of taking all 3 names. On swarm the ping for individual instance is not supported today is what I understand it as based on https://github.com/moby/moby/issues/30546

If I have the above solution, in my yaml compose file, I can just set the replica to 3 for the mongo service block and I know I can access the services via above named entries. Is this possible? Without this, I am creating 3 separate service blocks to achieve the same.

**Comment by azzeddinefaik** [comment]
I have here the same issue. 
I have a setup of docker swarm cluster using ansible.  
and I have 3 services running [ mongo , mysql , app ]. I want to fill data into my db based on number of instance of app services. So I need to access mysql service from my node [dev-01].
 I can access it inside my app conatiner [using service name]. but I need it to outside  [without going throught docker ps && docker inspect commands ] ?

**Comment by michaelkrog** [comment]
I have the same issue with mongodb AND with rabbitmq.

I wonder what docker techs recommend for these kind of setups where specific hostnames for each instance is needed for the mongodb/rabbitmq/etc to function as a cluster in Docker Swarm mode.

**Comment by mrapczynski** [comment]
@michaelkrog Here is an example of what I do. I still write the Compose service definitions, and I define a named service for each node and launch it as a stack. Works well enough, and gives me the per-container hostname resolution that I need for `etcd` to work (I use it to store user sessions backing an open-source SSO identity provider).

I believe if the Swarm mode were tweaked to assign each container/task of a service a _resolvable_, and _predictable_... hostname, then I might be able to stop doing this and switch to a single service with replicas.

However, a separate and unrelated issue is persistent storage for clustered services. In the example below I assign each `etcd` some persistent storage on the host for the snapshot and WAL logs - good storage to back the cluster is pretty important. In cases like this, replicas have not worked out, and so the issue of container hostnames becomes less important.

```yaml
version: '3'

services:
  etcd0:
    deploy:
      placement:
        constraints:
          - node.hostname == docker-ucp-worker1
      restart_policy:
        condition: on-failure
        window: 60s
      resources:
        limits:
          memory: 1g
    environment:
      - ETCD_NAME=shib-etcd0
      - ETCD_HEARTBEAT_INTERVAL=5000
      - ETCD_ELECTION_TIMEOUT=25000
      - ETCD_ADVERTISE_CLIENT_URLS=http://etcd0:2379,http://etcd0:4001
      - ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379,http://0.0.0.0:4001
      - ETCD_LISTEN_PEER_URLS=http://0.0.0.0:2380
      - ETCD_INITIAL_ADVERTISE_PEER_URLS=http://etcd0:2380
      - ETCD_INITIAL_CLUSTER_TOKEN=...
      - ETCD_INITIAL_CLUSTER=shib-etcd0=http://etcd0:2380,shib-etcd1=http://etcd1:2380
      - ETCD_INITIAL_CLUSTER_STATE=new
    logging:
      driver: json-file
    hostname: etcd0
    image: quay.io/coreos/etcd:v3.1.6
    networks:
      - shibboleth_test
    volumes:
      - /docker/shibboleth_test_etcd0_data:/shib-etcd0.etcd

  etcd1:
    deploy:
      placement:
        constraints:
          - node.hostname == docker-ucp-worker2
      restart_policy:
        condition: on-failure
        window: 60s
      resources:
        limits:
          memory: 1g
    environment:
      - ETCD_NAME=shib-etcd1
      - ETCD_HEARTBEAT_INTERVAL=5000
      - ETCD_ELECTION_TIMEOUT=25000
      - ETCD_ADVERTISE_CLIENT_URLS=http://etcd1:2379,http://etcd1:4001
      - ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379,http://0.0.0.0:4001
      - ETCD_LISTEN_PEER_URLS=http://0.0.0.0:2380
      - ETCD_INITIAL_ADVERTISE_PEER_URLS=http://etcd1:2380
      - ETCD_INITIAL_CLUSTER_TOKEN=...
      - ETCD_INITIAL_CLUSTER=shib-etcd0=http://etcd0:2380,shib-etcd1=http://etcd1:2380
      - ETCD_INITIAL_CLUSTER_STATE=new
    hostname: etcd1
    image: quay.io/coreos/etcd:v3.1.6
    logging:
      driver: json-file
    networks:
      - shibboleth_test
    volumes:
      - /docker/shibboleth_test_etcd1_data:/shib-etcd1.etcd

networks:
  shibboleth_test:
    external: true
```

**Comment by deitch** [comment]
I think people are asking for the equivalent of kubernetes' [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)

It need not be as complex and full-featured, but the predictability around it is beneficial for anything that has some sense of consistency and state. I use it for databases, etcd, Kafka, zookeeper and redis, just as a start. The overhead without it would have been quite large.

**Comment by thaJeztah** [comment]
An initial proposal looks to be here; https://github.com/docker/swarmkit/issues/1242, feedback/input welcome 👍

**Comment by deitch** [comment]
> feedback/input welcome

OK, I did.

**Comment by kleinsasserm** [comment]
I would like to drop in docker/libnetwork#1855 here, as I already tested this. And all of written here, I also mentioned in the PR, and I've also used nearly the same explanation. :)

**Comment by nicodmf** [comment]
closed in favor docker/swarmkit#1242
